code/modules/xenoarchaeology/xenoartifact_traits.dm

```diff
@@ -0,0 +1,1011 @@
+	..()
+
+/datum/xenoartifact_trait/activator/clock/calculate_charge(datum/source, obj/item/thing, mob/user, atom/target, params)
+	var/obj/item/xenoartifact/X = source
+	X.process_type = PROCESS_TYPE_TICK
+	START_PROCESSING(SSobj, X)
+	log_game("[user]:[isliving(user) ? user?.ckey : "no ckey"] set clock on [X] at [world.time] using [thing]. [X] located at [X.x] [X.y] [X.z].")
+
+/datum/xenoartifact_trait/activator/signal
+	label_name = "Signal"
+	label_desc = "Signal: The material recieves radio frequencies and reacts when a matching code is delivered."
+	charge = 25
+	signals = list(XENOA_SIGNAL)
+
+/datum/xenoartifact_trait/activator/signal/on_init(obj/item/xenoartifact/X)
+	..()
+	X.code = rand(1, 100) //Random code is shared by all signaller traits
+	X.frequency = FREQ_SIGNALER
+	X.set_frequency(X.frequency)
+	X.max_range += 1
+
+/datum/xenoartifact_trait/activator/signal/on_item(obj/item/xenoartifact/X, atom/user, atom/item)
+	if(istype(item, /obj/item/analyzer))
+		to_chat(user, "<span class='info'>The [item.name] displays a signal-input code of [X.code], and frequency [X.frequency].</span>")
+		return TRUE
+	..()
+
+/datum/xenoartifact_trait/activator/signal/calculate_charge(datum/source, obj/item/thing, mob/user, atom/target, params)
+	var/obj/item/xenoartifact/X = source
+	X.default_activate(charge, user, target)
+	log_game("[user]:[isliving(user) ? user?.ckey : "no ckey"] signalled [X] at [world.time]. [X] located at [X.x] [X.y] [X.z].")
+
+/datum/xenoartifact_trait/activator/batteryneed
+	desc = "Charged"
+	label_desc = "Charged: The material has a natural power draw. Supplying any current to this will cause a reaction."
+	charge = 25
+	signals = list(COMSIG_PARENT_ATTACKBY)
+
+/datum/xenoartifact_trait/activator/batteryneed/on_item(obj/item/xenoartifact/X, atom/user, atom/item)
+	if(istype(item, /obj/item/multitool))
+		to_chat(user, "<span class='info'>The [item.name] displays a draw of [X.charge_req].</span>")
+		return TRUE
+	..()
+
+/datum/xenoartifact_trait/activator/batteryneed/calculate_charge(datum/source, obj/item/thing, mob/user, atom/target, params)
+	var/obj/item/xenoartifact/X = source
+	if(istype(thing, /obj/item/stock_parts/cell))
+		var/obj/item/stock_parts/cell/C = thing
+		if(X.default_activate(charge, user, user))
+			C.charge -= X.charge_req*10
+
+//Minor traits - Use the to define aspects of the artifact without any immediate interaction
+/datum/xenoartifact_trait/minor/looped
+	desc = "Looped"
+	label_desc = "Looped: The Artifact feeds into itself and amplifies its own charge."
+
+/datum/xenoartifact_trait/minor/looped/on_item(obj/item/xenoartifact/X, atom/user, atom/item)
+	if(istype(item, /obj/item/multitool))
+		to_chat(user, "<span class='info'>The [item.name] displays a resistance reading of [X.charge_req*0.1].</span>") 
+		return TRUE
+	..()
+
+/datum/xenoartifact_trait/minor/looped/activate(obj/item/xenoartifact/X)
+	X.charge = ((100-X.charge)*0.2)+X.charge //This should generally cut off around 100
```

Comment:
Can this go over 100 or below 0?

Reply:
No, ``X.charge`` has an upper limit of 100.